src/VectorStore.ts

```diff
@@ -1,186 +1,292 @@
 	}
 
-	private vault: Vault
-	private dbFileName = "database2.json"
-	private dbFilePath = `.obsidian/plugins/vault-chat/${this.dbFileName}`
-
-	private filePathToVector: Map<string, Vector>;
-	isReady: Promise<boolean>;
-	isIndexingComplete: Promise<boolean>;
-
-	/*
-	   this is the most conservative rate limit for openai
-	   we are throttling this function to ensure indexing the vault will succeed
-	   this is an MVP solution - ideal solution would handle the rate limiting of a specific token since
-	   the rate limit could change over time and different tokens can have different rate limits
-	 */
-	private throttler = new Throttler({
-		tokensPerInterval: 20,
-		interval: "minute"
-	})
-
-	private async initializeFile() {
-		const fileSystemAdapter = this.vault.adapter
-		this.isReady = new Promise(async resolve => {
-			const vectorFileExists = await fileSystemAdapter.exists(this.dbFilePath)
-			if (vectorFileExists) {
-				const vectorFileContents = await fileSystemAdapter.read(this.dbFilePath)
-				try {
-					this.filePathToVector = new Map(JSON.parse(vectorFileContents))
-				} catch (e) {
-					console.error(`Unable to load existing database json, starting fresh`, e)
-					this.filePathToVector = new Map()
-					await this.saveVectorFile()
+	async initDatabase() {
+		this.embeddings = await this.readEmbeddingsFromDatabaseFile()
+	}
+
+	async updateDatabase(latestFiles: TFile[]) {
+		const newEmbeddings: Map<string, FileEntry> = new Map()
+		const oldEmbeddings: Map<string, FileEntry> = new Map(
+			JSON.parse(
+				JSON.stringify(Array.from(this.embeddings))
+			)
+		);
+		const entriesToUpdate: FileEntryUpdate[] = []
+		for (const file of latestFiles) {
+			const oldFileEntry = oldEmbeddings.get(file.path)
+			const fileContents = await this.vault.read(file)
+			const newHash = this.generateMd5Hash(fileContents)
+			if ((oldFileEntry && newHash !== oldFileEntry.md5hash) || // EXISTING FILE IN DB TO BE UPDATED
+				!oldFileEntry) { // NEW FILE IN DB TO BE ADDED
+				entriesToUpdate.push({
+					path: file.path,
+					chunk: false,
+					hash: newHash,
+					contents: fileContents,
+					embedding: undefined,
+					mTs: file.stat.mtime
+				})
+				const chunks = await this.chunkFile(fileContents) // todo handle empty files or very short files!
+				if (!chunks) {
+					console.error('failed to get chunks from file - file entry failing - lets skip it for now')
+					// todo should maybe track failed files
+					continue
 				}
-				resolve(true)
-			} else {
-				this.filePathToVector = new Map()
-				await this.saveVectorFile()
-				resolve(true)
+				chunks.forEach(chunk => {
+					entriesToUpdate.push({
+						path: file.path,
+						chunk: true,
+						hash: undefined,
+						contents: chunk,
+						embedding: undefined,
+						mTs: undefined
+					})
+				})
 			}
+		}
+		const embeddingRequestTexts = entriesToUpdate.map(e => e.contents)
+		const response = await this.createEmbeddingBatch(embeddingRequestTexts)
```

Comment:
Should probably batch these rather than letting it be infinitely large. 100 texts at a time? 500? 
In our vault with 600 files, even if each file had 4 chunks, that would be 3000 in one request.